blog/zh/blog/2023/03/09/release-apache-apisix-3.1.0.md

```diff
@@ -0,0 +1,227 @@
+            "variables": [
+                "name",
+                "githubAccount"
+            ]
+        }
+    }
+}'
+```
+
+这里 `query` 是我们要用到的查询语句，`variables` 是事先声明的变量列表。
+
+接下来就能像 RESTful 接口一样访问它：
+
+```shell
+curl --location --request POST 'http://localhost:9080/graphql' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "name": "Niek",
+    "githubAccount": "npalm"
+}'
+```
+
+结果跟直接用对应的 GraphQL 语句访问上游是一样的：
+
+```json
+{
+  "data": {
+    "persons": [
+      {
+        "id": "7",
+        "name": "Niek",
+        "blog": "https://040code.github.io",
+        "githubAccount": "npalm",
+        "talks": [
+          {
+            "id": "19",
+            "title": "GraphQL - The Next API Language"
+          },
+          {
+            "id": "20",
+            "title": "Immutable Infrastructure"
+          }
+        ]
+      }
+    ]
+  }
+}
+```
+
+你同样能用 GET 请求来访问同样的接口，这时候参数就需要通过 query string 来传递：
+
+```
+curl 'http://localhost:9080/graphql?name=Niek&githubAccount=npalm'
+```
+
+## 新特性：支持在每个日志插件上设置日志格式
+
+在 3.2 版本，我们整理了 APISIX 现有的十多个 access 日志插件。现在每个插件都支持配置自定义日志格式：
+
+1. 在该插件的 plugin metadata 中定义全局的日志格式
+2. 在具体的路由规则上的该插件的配置中定义当前路由的日志格式
+
+以 `clickhouse-logger` 为例，下面是定义全局日志格式的做法：
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/plugin_metadata/clickhouse-logger \
+-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+    "log_format": {
+        "host": "$host",
+        "@timestamp": "$time_iso8601",
+        "client_ip": "$remote_addr"
+    }
+}'
+```
+
+下面则是当前路由的日志格式：
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 \
+-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+      "plugins": {
+            "clickhouse-logger": {
+                "log_format": {
+                    "host": "$host",
+                    "@timestamp": "$time_iso8601",
+                    "client_ip": "$remote_addr"
+                },
+                "user": "default",
+                "password": "a",
+                "database": "default",
+                "logtable": "test",
+                "endpoint_addrs": ["http://127.0.0.1:8123"]
+            }
+       },
+      "upstream": {
+           "type": "roundrobin",
+           "nodes": {
+               "127.0.0.1:1980": 1
+           }
+      },
+      "uri": "/hello"
+}'
+
+```
+
+## 新插件：请求体/响应体转换
+
+你是否在为如何把返回 XML 的古早上游服务介绍给只接受 JSON 的现代客户端而苦恼？3.2.0 版本新增的 `body-transformer` 插件开源解决这个问题。
+
+`body-transformer` 插件支持 JSON 和 XML 之间的互相转换。不过这并非是它唯一能做的事。它还支持通过模板配置输入输出的内容的具体格式。举个例子，
+
+假设你有下面的 JSON 模板：`{"foo":"{{name .. " world"}}","bar":{{age+10}}}`，并把它配置到 body-transformer 插件的 `request.template` 字段中：
+
+```
+      ...
+      "body-transformer": {
+          "request": {
+              "template": "..."
+          }
+      }
+      ...
```

Comment:
format

Reply:
"body-transformer": {
    "request": {
        "template": "..."
    }
}

like this?